### PR TITLE
Adds api to fetch related documents by tags

### PR DESCRIPTION
**Pull request by paoloromolini**
Adds a django rest list view to get related projects. 
This
/docsitalia/api/projects-by-tag/?tags=comuni,anagrafe
will return:

```
{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "id": 40,
            "name": "Matrimoni nel Comune di Tre Civette sul Comò",
            "slug": "matrimoni-delle-civette",
            "description": "Lorem ipsum dolor sit amet, consectetur \nadipisicing elit, sed do eiusmod tempor\nincididunt ut labore et dolore magna aliqua. \nUt enim ad minim veniam, quis nostrud \nexercitation ullamco laboris nisi ut \naliquip ex ea commodo consequat.\nDuis aute irure dolor in reprehenderit in \nvoluptate velit esse cillum dolore eu\nfugiat nulla pariatur. Excepteur sint \noccaecat cupidatat non proident, sunt in\nculpa qui officia deserunt mollit anim id \nest laborum.",
            "canonical_url": "http://localhost:8000/docs/matrimoni-delle-civette/en/bozza/",
            "publisher": {
                "name": "ANPR - Anagrafe Nazionale Popolazione Residente",
                "canonical_url": "http://localhost:8000/ANPR-Anagrafe-Nazionale-Popolazione-Residente/"
            },
            "publisher_project": {
                "canonica_url": "http://localhost:8000/ANPR-Anagrafe-Nazionale-Popolazione-Residente/lorem/",
                "name": "Lorem"
            }
        }
    ]
}
```

### CODE REVIEW

File: readthedocs/docsitalia/urls.py [hunk 1]
```diff
@@ -20,6 +24,7 @@
 
 
 urlpatterns = [
+    url(r'^docsitalia/', include(docsitalia_urls)),
```
Comment (paoloromolini) [inline readthedocs/docsitalia/urls.py:27]: I needed to move this up because otherwise a rtd test failed:
```
________________________ APIUnauthAccessTest.test_api_urls _________________________

self = <readthedocs.rtd_tests.tests.test_privacy_urls.APIUnauthAccessTest testMethod=test_api_urls>
get_public_task_data = <MagicMock name='get_public_task_data' id='139883792972240'>

    @mock.patch('readthedocs.restapi.views.task_views.get_public_task_data')
    def test_api_urls(self, get_public_task_data):
        from readthedocs.restapi.urls import urlpatterns
        get_public_task_data.side_effect = TaskNoPermission('Nope')
>       self._test_url(urlpatterns)

rtd_tests/tests/test_privacy_urls.py:374: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
rtd_tests/tests/test_privacy_urls.py:127: in _test_url
    self.assertResponse(path=path, name=name)
rtd_tests/tests/test_privacy_urls.py:83: in assertResponse
    value=resp_val))
E   AssertionError: Attribute mismatch for view api-root (/docsitalia/): status_code != 200 (got 404)

```
This was due the fact that the api root didn't worked

File: readthedocs/docsitalia/views/api.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+"""Docs italia api"""
+from __future__ import absolute_import
+
+from rest_framework import serializers, viewsets
+from rest_framework import mixins
+
+from readthedocs.projects.models import Project
+
+
+class ProjectsByTagSerializer(serializers.HyperlinkedModelSerializer):
```
Comment (xrmx) [inline readthedocs/docsitalia/views/api.py]: serializers go in docsitalia/serializers.py

File: readthedocs/docsitalia/views/api.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    serializer_class = ProjectsByTagSerializer
+    allowed_methods = ('GET')
+
+    def get_queryset(self):
+        """filter projects by tags"""
+        tags = self.request.query_params.get('tags', None)
+        if tags:
+            tags = tags.split(',')
+            queryset = Project.objects.filter(tags__name__in=tags).distinct()
```
Comment (xrmx) [inline readthedocs/docsitalia/views/api.py]: filter on slug and not name, no need for distinct, slugs should be unique

File: readthedocs/rtd_tests/tests/test_docsitalia.py [hunk 4]
```diff
@@ -537,3 +537,45 @@ def test_on_webhook_github_signal_ignores_invalid_branches(self):
...
+    def test_projects_by_tag_api(self):
+        project = Project.objects.create(
+            name='my project',
+            slug='myprojectslug',
+            repo='https://github.com/testorg/myrepourl.git'
+        )
+        data = {
+            'ref': 'master'
+        }
+        with requests_mock.Mocker() as rm:
```
Comment (xrmx) [inline readthedocs/rtd_tests/tests/test_docsitalia.py]: please just add plain tags to the model

Comment (xrmx) [reply readthedocs/rtd_tests/tests/test_docsitalia.py]: which is basically the next test, so just drop this

File: readthedocs/rtd_tests/tests/test_docsitalia.py [hunk 5]
```diff
@@ -537,3 +537,45 @@ def test_on_webhook_github_signal_ignores_invalid_branches(self):
...
+                'https://raw.githubusercontent.com/testorg/'
+                'myrepourl/master/document_settings.yml',
+                text=DOCUMENT_METADATA)
+            webhook_github.send(Project, project=project, data=data, event='push')
+        project.refresh_from_db()
+        response = self.client.get(reverse('projects-by-tag-list'), {'tags': 'amazing document'})
+        self.assertEqual(len(response.data['results']), 1)
+        self.assertEqual(response.status_code, 200)
+
+    def test_projects_by_tag_api_manual_add(self):
```
Comment (xrmx) [inline readthedocs/rtd_tests/tests/test_docsitalia.py]: est_projects_by_tag_api_filter_tags

File: readthedocs/rtd_tests/tests/test_docsitalia.py [hunk 6]
```diff
@@ -537,3 +537,45 @@ def test_on_webhook_github_signal_ignores_invalid_branches(self):
...
+            name='my project',
+            slug='myprojectslug',
+            repo='https://github.com/testorg/myrepourl.git'
+        )
+        project.tags.add('lorem', 'ipsum')
+        response = self.client.get(reverse('projects-by-tag-list'), {'tags': 'lorem, sicut'})
+        self.assertEqual(len(response.data['results']), 1)
+        self.assertEqual(response.status_code, 200)
+
+    def test_projects_by_tag_api_fail(self):
```
Comment (xrmx) [inline readthedocs/rtd_tests/tests/test_docsitalia.py]: Looks like test_projects_by_tag_returns_only_data_that_matches_tags or similar.
You are missing a test with no tags it returns an empty queryset

**Review by xrmx — CHANGES_REQUESTED** [review]

File: readthedocs/docsitalia/views/api.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    """
+    Projects by Tag DRF list model mixin
+
+    Takes a GET with tags separated by a comma:
+    tags=tag1,tag2
+    """
+
+    serializer_class = ProjectsByTagSerializer
+    allowed_methods = ('GET')
```
Comment (xrmx) [inline readthedocs/docsitalia/views/api.py]: forgot that this should be a tuple and not a string

File: readthedocs/docsitalia/views/api.py (hunk 3 above)
Comment (xrmx) [reply readthedocs/docsitalia/views/api.py]: ah the distinct is on the projects with multiple matched tags, sorry!

Comment (paoloromolini) [reply readthedocs/docsitalia/views/api.py]: I was following  taggit docs http://django-taggit.readthedocs.io/en/latest/api.html#filtering

Comment (xrmx) [reply readthedocs/docsitalia/views/api.py]: That's an example with plain orm though, we need something that won't be break things in urls :)

File: readthedocs/docsitalia/urls.py (hunk 1 above)
Comment (xrmx) [reply readthedocs/docsitalia/urls.py:27]: This test failure does not make any sense to me :)

Comment (paoloromolini) [reply readthedocs/docsitalia/urls.py:27]: Yes because the api root was treated as a non exisiting publisher if I remember correctly

File: readthedocs/docsitalia/serializers.py [hunk 8]
```diff
@@ -3,14 +3,37 @@
...
+    publisher = serializers.SerializerMethodField()
+    publisher_project = serializers.SerializerMethodField()
+
+    class Meta(ProjectSerializer.Meta):
+        fields = (
+            'id', 'name', 'slug', 'description',
+            'canonical_url', 'publisher', 'publisher_project',
+        )
+
+    @classmethod
```
Comment (xrmx) [inline readthedocs/docsitalia/serializers.py]: c'è un motivo del perchè questi siano classmethod?

File: readthedocs/docsitalia/models.py [hunk 9]
```diff
@@ -152,6 +153,11 @@ def get_absolute_url(self):
         """Get absolute url for publisher"""
         return reverse('publisher_detail', args=[self.slug])
 
+    def get_canonical_url(self):
+        """Get canonical url for publisher"""
+        production_domain = getattr(dj_settings, 'PRODUCTION_DOMAIN', None)
+        return 'http://{}{}'.format(production_domain, self.get_absolute_url())
```
Comment (xrmx) [inline readthedocs/docsitalia/models.py]: riusciamo a non hardcodare http?

File: readthedocs/docsitalia/serializers.py [hunk 10]
```diff
@@ -3,14 +3,37 @@
 from __future__ import absolute_import
 
 from rest_framework import serializers
+from readthedocs.restapi.serializers import ProjectSerializer
 
-from readthedocs.projects.models import Project
 
-
-class ProjectsByTagSerializer(serializers.HyperlinkedModelSerializer):
+class ProjectsByTagSerializer(ProjectSerializer):
```
Comment (xrmx) [inline readthedocs/docsitalia/serializers.py]: questa magari ha senso chiamarlo DocumentoSerializer o DocsItaliaProjectSerializer?

File: readthedocs/docsitalia/serializers.py [hunk 11]
```diff
@@ -3,14 +3,37 @@
...
     """Projects by Tag DRF Serializer"""
 
-    class Meta:
-        model = Project
-        fields = ('name', 'description', 'url')
+    publisher = serializers.SerializerMethodField()
+    publisher_project = serializers.SerializerMethodField()
+
+    class Meta(ProjectSerializer.Meta):
+        fields = (
```
Comment (xrmx) [inline readthedocs/docsitalia/serializers.py:17]: visto che ci siamo in un altro commit mettiamo pure i tags

File: readthedocs/docsitalia/models.py (hunk 9 above)
Comment (paoloromolini) [reply readthedocs/docsitalia/models.py]: Forse così https://github.com/rtfd/readthedocs.org/blob/f74d4fbe17ef7ba7fdd90616e75df84c4636c94c/readthedocs/core/resolver.py#L151

Comment (xrmx) [reply readthedocs/docsitalia/models.py]: si e noi abbiamo un docsitalia/resolver.py che possiamo estendere come ci pare

Comment (paoloromolini) [reply readthedocs/docsitalia/models.py]: perfetto

File: readthedocs/docsitalia/serializers.py (hunk 10 above)
Comment (paoloromolini) [reply readthedocs/docsitalia/serializers.py]: si la seconda mi piace

File: readthedocs/docsitalia/serializers.py (hunk 8 above)
Comment (paoloromolini) [reply readthedocs/docsitalia/serializers.py]: per pylint

File: readthedocs/docsitalia/serializers.py (hunk 11 above)
Comment (paoloromolini) [reply readthedocs/docsitalia/serializers.py:17]: Questa la facciamo nell' altra pr? Tanto è aperta su questa

Comment (xrmx) [reply readthedocs/docsitalia/serializers.py:17]: no perchè l'altra pr non serve più se metti i tags qua, nel senso che una volta che abbiamo il serializzatore lo possiamo usare nel ProjectViewSet che prende gli slug al posto delle pk

File: readthedocs/docsitalia/serializers.py (hunk 8 above)
Comment (paoloromolini) [reply readthedocs/docsitalia/serializers.py]: meglio uno staticmethod?

Comment (xrmx) [reply readthedocs/docsitalia/serializers.py]: obey the pylint god!

File: readthedocs/docsitalia/serializers.py (hunk 11 above)
Comment (paoloromolini) [reply readthedocs/docsitalia/serializers.py:17]: ok

Comment (paoloromolini) [reply readthedocs/docsitalia/serializers.py:17]: Il projectviewset va creato in questa pr perchè non c'è

Comment (xrmx) [reply readthedocs/docsitalia/serializers.py:17]: ce l'ho in pronto in un 'altra pr ma mi servono le modifiche hai fatto su questa

Comment (paoloromolini) [reply readthedocs/docsitalia/serializers.py:17]: allora aspetta che pusho :-)

File: readthedocs/docsitalia/resolver.py [hunk 12]
```diff
@@ -87,3 +87,26 @@ def resolve(self, project, protocol='http', filename='', private=None, **kwargs)
...
+    def resolve_docsitalia(self, project, publisher_slug, pb_project_slug=None, private=None, protocol='http'):
+        """
+        Resolve the complete URL fot a publisher or a publisher project
+
+        :param publisher_slug: the publisher slug
+        :param pb_project_slug: the publisher project slug
+        """
+        public_domain = getattr(settings, 'PUBLIC_DOMAIN', None)
+        use_https = getattr(settings, 'PUBLIC_DOMAIN_USES_HTTPS', True)
+        domain = self.resolve_domain(project, private=private)
```
Comment (xrmx) [inline readthedocs/docsitalia/resolver.py]: domain = getattr(settings, 'PRODUCTION_DOMAIN')

e possiamo droppare project come primo parametro

File: readthedocs/docsitalia/resolver.py [hunk 13]
```diff
@@ -87,3 +87,26 @@ def resolve(self, project, protocol='http', filename='', private=None, **kwargs)
...
+        if use_https and public_domain and public_domain in domain:
+            protocol = 'https'
+
+        if pb_project_slug:
+            path = '{}/{}'.format(publisher_slug, pb_project_slug)
```
Comment (xrmx) [inline readthedocs/docsitalia/resolver.py]: stringhe unicode

File: readthedocs/docsitalia/resolver.py [hunk 14]
```diff
@@ -87,3 +87,26 @@ def resolve(self, project, protocol='http', filename='', private=None, **kwargs)
...
+        else:
+            path = publisher_slug
+        return '{}://{}/{}'.format(
```
Comment (xrmx) [inline readthedocs/docsitalia/resolver.py]: pure qua

File: readthedocs/docsitalia/resolver.py [hunk 15]
```diff
@@ -87,3 +87,26 @@ def resolve(self, project, protocol='http', filename='', private=None, **kwargs)
...
+        Resolve the complete URL fot a publisher or a publisher project
+
+        :param publisher_slug: the publisher slug
+        :param pb_project_slug: the publisher project slug
+        """
+        public_domain = getattr(settings, 'PUBLIC_DOMAIN', None)
+        use_https = getattr(settings, 'PUBLIC_DOMAIN_USES_HTTPS', True)
+        domain = self.resolve_domain(project, private=private)
+        if use_https and public_domain and public_domain in domain:
+            protocol = 'https'
```
Comment (xrmx) [inline readthedocs/docsitalia/resolver.py]: u'https'

File: readthedocs/docsitalia/serializers.py [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+            'canonical_url', 'publisher', 'publisher_project',
+        )
+
+    @staticmethod
+    def get_publisher(obj):
+        """gets the publisher"""
+        p_p = obj.publisherproject_set.first()
+        if p_p.publisher:
+            return {
+                'name': p_p.publisher.name,
```
Comment (xrmx) [inline readthedocs/docsitalia/serializers.py]: non devi usare il name del publisher ma la roba che c'è nel metadata

**Review by xrmx — CHANGES_REQUESTED** [review]

File: readthedocs/docsitalia/resolver.py (hunk 12 above)
Comment (paoloromolini) [reply readthedocs/docsitalia/resolver.py]: ok quindi come avevo messo all'inizio

File: readthedocs/docsitalia/models.py [hunk 17]
```diff
@@ -191,6 +195,11 @@ def get_absolute_url(self):
         """get absolute url for publisher project"""
         return reverse('publisher_project_detail', args=[self.publisher.slug, self.slug])
 
+    def get_canonical_url(self):
+        """get canonical url for publisher project"""
+        project = self.projects.first()
```
Comment (xrmx) [inline readthedocs/docsitalia/models.py]: questo non ti serve più

**Review by xrmx — APPROVED** [review]
a parte la riga in più il resto va bene

**Comment by paoloromolini** [comment]
C'è un problema, non possiamo fare l'override dei settings di CLASS_OVERRIDES perchè non funziona. Con @yakky abbiamo pensato di creare un altro env di tox nel quale far girare solo i test marked come ItalaResolverSettings che utilizzerano un particolare settings che carica la class_overrides e skipparli invece negli altri env.

**Comment by xrmx** [comment]
oppure aggiorniamo i test per passare con la conf di default e aggiungiamo un test che testa il metodo del nostro resolver separatamente.

**Comment by yakky** [comment]
@xrmx non sono convinto che testare solo questo metodo https://github.com/italia/docs.italia.it/pull/30/files#diff-7df2d59eef3d43a7be3f4ebf600baed8R91 sia significativo
Senza poter verificare che venga effettivamente usato nei serializer potrebbe avere poco senso, non credi?
Dal momento che stiamo introducendo metodi specifici che però funzionano solo con il resolver custom, non vedo come si riesca a chiudere il giro, diversamente

**Comment by xrmx** [comment]
Se nel serializer testo che chiamo la funzione (mock) e se testando la funziona con gli stessi parametri funziona per me il giro è chiuso. Detto questo fate come vi pare, a me basta che la roba sia testata, che i test passino e che sta PR sia mergiata :)

**Comment by paoloromolini** [comment]
@rm @yakky pushato :smiley:

**Comment by xrmx** [comment]
squasha un po' di commit, che mergiamo
